flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskIOSchedulerImpl.java

```diff
@@ -0,0 +1,426 @@
+                    && numRequestedBuffers < bufferPool.getAverageBuffersPerRequester()) {
+                isRunning = true;
+                ioExecutor.execute(
+                        () -> {
+                            try {
+                                run();
+                            } catch (Throwable throwable) {
+                                LOG.error("Failed to read data.", throwable);
+                                // handle un-expected exception as unhandledExceptionHandler is not
+                                // worked for ScheduledExecutorService.
+                                FatalExitExceptionHandler.INSTANCE.uncaughtException(
+                                        Thread.currentThread(), throwable);
+                            }
+                        });
+            }
+        }
+    }
+
+    private long getBufferRequestTimeoutTime() {
+        return bufferPool.getLastBufferOperationTimestamp() + bufferRequestTimeout.toMillis();
+    }
+
+    /**
+     * The {@link ScheduledSubpartitionReader} is responsible for reading a subpartition from disk,
+     * and is scheduled by the {@link DiskIOScheduler}.
+     */
+    private class ScheduledSubpartitionReader implements Comparable<ScheduledSubpartitionReader> {
+
+        private final TieredStorageSubpartitionId subpartitionId;
+
+        private final NettyConnectionWriter nettyConnectionWriter;
+
+        private int nextBufferIndex = 0;
+
+        private boolean isFailed;
+
+        public ScheduledSubpartitionReader(
+                TieredStorageSubpartitionId subpartitionId,
+                NettyConnectionWriter nettyConnectionWriter) {
+            this.subpartitionId = subpartitionId;
+            this.nettyConnectionWriter = nettyConnectionWriter;
+        }
+
+        public void loadDiskDataToBuffers(Queue<MemorySegment> buffers, BufferRecycler recycler)
+                throws IOException {
+
+            if (isFailed) {
+                throw new IOException(
+                        "The scheduled subpartition reader for "
+                                + subpartitionId
+                                + " has already been failed.");
+            }
+
+            // If the number of written but unsent buffers achieves the limited value, skip this
+            // time.
+            if (nettyConnectionWriter.numQueuedBuffers() >= maxBufferReadAhead) {
+                return;
+            }
+            while (!buffers.isEmpty()
+                    && nettyConnectionWriter.numQueuedBuffers() < maxBufferReadAhead) {
+                MemorySegment memorySegment = buffers.poll();
+                Buffer buffer;
+                try {
+                    if ((buffer =
+                                    partitionFileReader.readBuffer(
+                                            partitionId,
+                                            subpartitionId,
+                                            -1,
```

Comment:
We should not assume the partition file reader would ignore the segment id.

Reply:
Fixed.